src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -128,6 +133,97 @@ internal partial class Binder
+                ? BindInitializerExpression(syntax: node.Initializer, type: strippedType, typeSyntax: syntax, diagnostics)
+                : null;
+
+            BoundExpression operand = source;
+            bool hasErrors;
+            switch (strippedType.TypeKind)
+            {
+                case TypeKind.Array:
+                case TypeKind.Enum:
+                case TypeKind.Delegate:
+                case TypeKind.Interface:
+                    Error(diagnostics, ErrorCode.ERR_BadTargetTypeForNew, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Pointer:
+                    Error(diagnostics, ErrorCode.ERR_UnsafeTypeInObjectCreation, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Dynamic:
+                    Error(diagnostics, ErrorCode.ERR_NoConstructors, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Struct when strippedType.IsTupleType:
+                    Error(diagnostics, ErrorCode.ERR_NewWithTupleTypeSyntax, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Struct:
+                case TypeKind.Class:
+                    operand = BindClassCreationExpression(
+                        node,
+                        typeName: strippedType.Name,
+                        typeNode: node,
+                        type: (NamedTypeSymbol)strippedType,
+                        arguments,
+                        diagnostics,
+                        boundInitializerOpt,
+                        forTargetTypedNew: true);
+                    operand.WasCompilerGenerated = true;
+                    hasErrors = operand.HasErrors;
+                    break;
+
+                case TypeKind.TypeParameter:
+                    var typeParameter = (TypeParameterSymbol)strippedType;
+                    if (!typeParameter.IsInstantiable())
+                    {
+                        goto case TypeKind.Array;
+                    }
+
+                    operand = BindTypeParameterCreationExpression(
+                        node,
+                        typeParameter,
+                        boundInitializerOpt,
+                        diagnostics);
+                    operand.WasCompilerGenerated = true;
+                    hasErrors = operand.HasErrors;
+                    break;
+
+                default:
+                    hasErrors = true;
+                    break;
+            }
+
+            arguments.Free();
+
+            return new BoundConversion(
+                syntax,
+                operand,
+                conversion: destination.IsNullableType()
+                    ? new Conversion(ConversionKind.ImplicitNullable, ImmutableArray.Create(conversion))
```

Comment:
Is this the right approach or it's better to nest two BoundConversion for easier check in `Conversion.IsNew`?

Reply:
It looks like other places nest BoundConversions, rather than tweaking the Conversion itself. Let's do that here too.

---
In reply to: [202642878](https://github.com/dotnet/roslyn/pull/25196#discussion_r202642878) [](ancestors = 202642878)